### PR TITLE
Call torch.use_deterministic_algorithms(True) for tests that require determinism

### PR DESCRIPTION
**Pull request by fxmarty**
Although it appears that on Nvidia GPUs all model tests requiring determinism pass without this option, this is not the case on AMD MI210, and it is not safe to assume in general. For example `CUDA_VISIBLE_DEVICES=1 pytest tests/models/cpmant/test_modeling_cpmant.py -s -k "test_determinism"` do not pass on main but do pass with this PR. Looking at the intermediate logits, it appears that a nn.Linear call is responsible of the deviation.

### CODE REVIEW

**Review by amyeroberts — COMMENTED** [review]
Thanks for investigating and adding this @fxmarty!

My only concern at the moment is that this is changing global state, which means that tests' outputs might become dependent on which tests are executed first. Is it possible to handle this within a context manager, or maybe just set at the top of the module? @ydshieh what do you think?

**Comment by fxmarty** [comment]
> First question: do not pass on main --> I think you are talking about running on AMD GPU?

Yes, on MI210 with RoCm 5.4.2.

> Could we wait until all the AMD GPU CI runner is set up, and let's run it without any modification to see what fail and what pass. Then we can start to working on failing tests.

It is fine, although by running manually the transformers tests it appears that `CUDA_VISIBLE_DEVICES=0 pytest tests/models/cpmant/test_modeling_cpmant.py -s -k "test_determinism"` do not pass.

@ydshieh Yes, I did not think about the RuntimeError. Would it make sense to run these tests under `torch.use_deterministic_algorithms(True)`, catch relevant `RuntimeError` and if happening, fall back on `torch.use_deterministic_algorithms(False)`?

**Comment by fxmarty** [comment]
Actually, we can use `warn_only=True`.

File: tests/test_modeling_common.py [hunk 1]
```diff
@@ -224,6 +224,7 @@ def _prepare_for_class(self, inputs_dict, model_class, return_labels=False):
 
     def test_save_load(self):
```
Comment (ArthurZucker) [inline tests/test_modeling_common.py:225]: would it make sense to have a `decorator` for this? which would set before and unset after?

Comment (fxmarty) [reply tests/test_modeling_common.py:225]: sure!

Comment (fxmarty) [reply tests/test_modeling_common.py:227]: Thank you, this is addressed.

**Review by ArthurZucker — COMMENTED** [review]
Changes look good, let's just make the tests go green!

**Comment by ydshieh** [comment]
From my side, I would love to have this decorator applied to the test methods that are implemented for the relevant model test classes (here `CpmAntModelTest`), i.e. override from the common ones.

WDYT?
